[GStreamer] Video not resized with playbin3 on i.MX8M Plus and COG

marex opened the pull request:
#### 1bcae4eccfb945436e2c9b923c3293d94548c80b
<pre>
[GStreamer] Video not resized with playbin3 on i.MX8M Plus and COG
<a href="https://bugs.webkit.org/show_bug.cgi?id=263165">https://bugs.webkit.org/show_bug.cgi?id=263165</a>

Reviewed by Philippe Normand.

When using WEBKIT_GST_USE_PLAYBIN3=1 and cog to open VP9 video from
remote server which provides a video with html5 video tag:

&quot;
&lt;video id=&quot;video&quot; autoplay muted loop autofocus controls&gt;
&lt;source src=&quot;./1.webm&quot; type=&quot;video/webm&quot;&gt;
&lt;/video&gt;
&quot;

then only single message is delivered
to MediaPlayerPrivateGStreamer::handleStreamCollectionMessage() . The
current workaround check would bail out since the message source is
decodebin3, while the m_source is webkitwebsrc . This prevents
player-&gt;updateTracks from being called, and thus m_hasVideo from being
set, and thus hasVideo() in MediaPlayerPrivateGStreamer::naturalSize()
returns false and MediaPlayerPrivateGStreamer::naturalSize returns
FloatSize() size, which is 0x0. The resulting video element in the
browser has minimum height set and is not correctly resized.

Limit the MediaPlayerPrivateGStreamer::handleStreamCollectionMessage()
workaround only to non-filesrc and non-webkitwebsrc, since webkitwebsrc
is not generating the stream-collection events.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleStreamCollectionMessage):

Canonical link: <a href="https://commits.webkit.org/269347@main">https://commits.webkit.org/269347@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2cd7c5fb5a6e3606362a7632af1af6c1dd20ce5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23374 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24199 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20637 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22549 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26760 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22829 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21649 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22535 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19351 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25055 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19286 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20207 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26458 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20303 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20438 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24318 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20968 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20426 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5309 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24656 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20964 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1554,7 +1554,7 @@ void MediaPlayerPrivateGStreamer::handleStreamCollectionMessage(GstMessage* mess
     // and sometimes with duplicated streams. Let's only listen for stream-collection messages from
     // the source to avoid these issues.
-    if (!g_str_has_prefix(GST_OBJECT_NAME(m_source.get()), "filesrc") && GST_MESSAGE_SRC(message) != GST_OBJECT(m_source.get())) {
+    if (!(g_str_has_prefix(GST_OBJECT_NAME(m_source.get()), "filesrc") || g_str_has_prefix(GST_OBJECT_NAME(m_source.get()), "webkitwebsrc")) && GST_MESSAGE_SRC(message) != GST_OBJECT(m_source.get())) {
```
`WEBKIT_IS_WEB_SRC(m_source.get()) `

Commit:
[GStreamer] Video not resized with playbin3 on i.MX8M Plus and COG
https://bugs.webkit.org/show_bug.cgi?id=263165

Reviewed by Philippe Normand.

When using WEBKIT_GST_USE_PLAYBIN3=1 and cog to open VP9 video from
remote server which provides a video with html5 video tag:

"
<video id="video" autoplay muted loop autofocus controls>
<source src="./1.webm" type="video/webm">
</video>
"

then only single message is delivered
to MediaPlayerPrivateGStreamer::handleStreamCollectionMessage() . The
current workaround check would bail out since the message source is
decodebin3, while the m_source is webkitwebsrc . This prevents
player->updateTracks from being called, and thus m_hasVideo from being
set, and thus hasVideo() in MediaPlayerPrivateGStreamer::naturalSize()
returns false and MediaPlayerPrivateGStreamer::naturalSize returns
FloatSize() size, which is 0x0. The resulting video element in the
browser has minimum height set and is not correctly resized.

Limit the MediaPlayerPrivateGStreamer::handleStreamCollectionMessage()
workaround only to non-filesrc and non-webkitwebsrc, since webkitwebsrc
is not generating the stream-collection events.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleStreamCollectionMessage):

Canonical link: https://commits.webkit.org/269347@main
